### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

# Python City Schools Project

## Import libraries 
Pandas and numpy were utilised for this project

In [1]:
# Dependencies and Setup
import pandas as pd
import numpy as np

## Data Loading 

In [2]:
# File to Load (Remember to Change These)
school_data_to_load = "Resources/schools_complete.csv"
student_data_to_load = "Resources/students_complete.csv"

# Read School and Student Data File and store into Pandas DataFrames
school_data = pd.read_csv(school_data_to_load)
student_data = pd.read_csv(student_data_to_load)

In [3]:
school_data.head()

School ID            school_name      type  size   budget
0          0      Huang High School  District  2917  1910635
1          1   Figueroa High School  District  2949  1884411
2          2    Shelton High School   Charter  1761  1056600
3          3  Hernandez High School  District  4635  3022020
4          4    Griffin High School   Charter  1468   917500

In [4]:
student_data.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  
0             66          79  
1             94          61  
2             90          60  
3             67          58  
4             97          84

In [5]:
# Combine the data into a single dataset.  
school_data_complete = pd.merge(student_data, school_data, how="left", on="school_name")

In [6]:
school_data_complete

Student ID       student_name gender grade         school_name  \
0               0       Paul Bradley      M   9th   Huang High School   
1               1       Victor Smith      M  12th   Huang High School   
2               2    Kevin Rodriguez      M  12th   Huang High School   
3               3  Dr. Richard Scott      M  12th   Huang High School   
4               4         Bonnie Ray      F   9th   Huang High School   
...           ...                ...    ...   ...                 ...   
39165       39165       Donna Howard      F  12th  Thomas High School   
39166       39166          Dawn Bell      F  10th  Thomas High School   
39167       39167     Rebecca Tanner      F   9th  Thomas High School   
39168       39168       Desiree Kidd      F  10th  Thomas High School   
39169       39169    Carolyn Jackson      F  11th  Thomas High School   

       reading_score  math_score  School ID      type  size   budget  
0                 66          79          0  District  2917  1910635  
1                 94          61          0  District  2917  1910635  
2                 90          60          0  District  2917  1910635  
3                 67          58          0  District  2917  1910635  
4                 97          84          0  District  2917  1910635  
...              ...         ...        ...       ...   ...      ...  
39165             99          90         14   Charter  1635  1043130  
39166             95          70         14   Charter  1635  1043130  
39167             73          84         14   Charter  1635  1043130  
39168             99          90         14   Charter  1635  1043130  
39169             95          75         14   Charter  1635  1043130  

[39170 rows x 11 columns]

## Data Exploration 

## District Summary

* Calculate the total number of schools

* Calculate the total number of students

* Calculate the total budget

* Calculate the average math score 

* Calculate the average reading score

* Calculate the percentage of students with a passing math score (70 or greater)

* Calculate the percentage of students with a passing reading score (70 or greater)

* Calculate the percentage of students who passed math **and** reading (% Overall Passing)

* Create a dataframe to hold the above results

* Optional: give the displayed data cleaner formatting

### Total number of schools
There are 15 schools in total in the dataset provided for this project

In [7]:
#district summary

total_schools = school_data_complete["school_name"].nunique()
student_sum = school_data_complete["Student ID"].nunique()
total_budget = "${:,.2f}".format(school_data["budget"].sum())

average_maths_score = round((school_data_complete["math_score"].mean()), 2)
average_read_score = round((school_data_complete["reading_score"].mean()), 2)

goodmathsers = school_data_complete[school_data_complete["math_score"] >= 70]
pass_maths = round((len(goodmathsers)/student_sum * 100), 2)

goodreaders = school_data_complete[school_data_complete["reading_score"] >= 70]
pass_read = round((len(goodreaders)/student_sum * 100), 2)

goodboth = school_data_complete[(school_data_complete["reading_score"] >= 70) & (school_data_complete["math_score"] >= 70)]
pass_both = round((len(goodboth)/student_sum * 100), 2)

district_df = pd.DataFrame({
    "Total Schools": [total_schools],
    "Total Students": [student_sum],
    "Total Budget": [total_budget], 
    "Average Maths Score": [average_maths_score],
    "Average Reading Score": [average_read_score],
    "% Passing Maths": [pass_maths],
    "% Passing Reading": [pass_read],
    "% Overall Passing": [pass_both],
})
district_df

Total Schools  Total Students    Total Budget  Average Maths Score  \
0             15           39170  $24,649,428.00                78.99   

   Average Reading Score  % Passing Maths  % Passing Reading  \
0                  81.88            74.98              85.81   

   % Overall Passing  
0              65.17

## School Summary

* Create an overview table that summarizes key metrics about each school, including:
  * School Name
  * School Type
  * Total Students
  * Total School Budget
  * Per Student Budget
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * % Overall Passing (The percentage of students that passed math **and** reading.)
  
* Create a dataframe to hold the above results

In [8]:
x = school_data_complete.groupby("school_name")
y = x["reading_score"].mean()
z= y.mean()
z

82.52918801528318

In [9]:
#school summary

school_sum = school_data_complete.groupby("school_name").agg({
                                                                "type": lambda column: column.unique(), 
                                                                "size": lambda column: column.unique(), 
                                                                "budget": lambda column: column.unique(),
                                                                "reading_score": "mean", 
                                                                "math_score": "mean"})
#means = school_data_complete.groupby(["school_name"]).mean()
#school_sum = means[["size", "budget", "math_score", "reading_score"]]

#rename the existing columns
school_sum = school_sum.rename(columns={
                                "type": "Type",
                                "size": "Total Students",
                                "budget": "Total School Budget",
                                "math_score": "Average Math Score",
                                "reading_score": "Average Reading Score"})

#calculate averages
school_sum["Per Student Budget"] = (school_sum["Total School Budget"]/school_sum["Total Students"])

x = goodmathsers.groupby(["school_name"])
school_sum["% Passing Math"] = round((x.count()["math_score"]/school_sum["Total Students"] * 100), 2)

y = goodreaders.groupby(["school_name"])
school_sum["% Passing Reading"] = round((y.count()["reading_score"]/school_sum["Total Students"] * 100), 2)

z = goodboth.groupby(["school_name"])
school_sum["% Overall Passing"] = round((z.count()["reading_score"]/school_sum["Total Students"] * 100), 2)



#change the order of columns:
school_sum = school_sum[[
      "Type",   
      "Total Students", 
      "Total School Budget", 
      "Per Student Budget", 
      "Average Math Score", 
      "Average Reading Score",
      "% Passing Math",
      "% Passing Reading",
      "% Overall Passing"
     ]]

#format numbers nicely
school_sum_dollar = round(school_sum, 2)
school_sum_dollar["Total School Budget"] = school_sum_dollar["Total School Budget"].map("${:,.2f}".format)
school_sum_dollar["Per Student Budget"] = school_sum_dollar["Per Student Budget"].map("${:,.2f}".format)
school_sum_dollar

Type  Total Students Total School Budget  \
school_name                                                           
Bailey High School     District            4976       $3,124,928.00   
Cabrera High School     Charter            1858       $1,081,356.00   
Figueroa High School   District            2949       $1,884,411.00   
Ford High School       District            2739       $1,763,916.00   
Griffin High School     Charter            1468         $917,500.00   
Hernandez High School  District            4635       $3,022,020.00   
Holden High School      Charter             427         $248,087.00   
Huang High School      District            2917       $1,910,635.00   
Johnson High School    District            4761       $3,094,650.00   
Pena High School        Charter             962         $585,858.00   
Rodriguez High School  District            3999       $2,547,363.00   
Shelton High School     Charter            1761       $1,056,600.00   
Thomas High School      Charter            1635       $1,043,130.00   
Wilson High School      Charter            2283       $1,319,574.00   
Wright High School      Charter            1800       $1,049,400.00   

                      Per Student Budget  Average Math Score  \
school_name                                                    
Bailey High School               $628.00               77.05   
Cabrera High School              $582.00               83.06   
Figueroa High School             $639.00               76.71   
Ford High School                 $644.00               77.10   
Griffin High School              $625.00               83.35   
Hernandez High School            $652.00               77.29   
Holden High School               $581.00               83.80   
Huang High School                $655.00               76.63   
Johnson High School              $650.00               77.07   
Pena High School                 $609.00               83.84   
Rodriguez High School            $637.00               76.84   
Shelton High School              $600.00               83.36   
Thomas High School               $638.00               83.42   
Wilson High School               $578.00               83.27   
Wright High School               $583.00               83.68   

                       Average Reading Score  % Passing Math  \
school_name                                                    
Bailey High School                     81.03           66.68   
Cabrera High School                    83.98           94.13   
Figueroa High School                   81.16           65.99   
Ford High School                       80.75           68.31   
Griffin High School                    83.82           93.39   
Hernandez High School                  80.93           66.75   
Holden High School                     83.81           92.51   
Huang High School                      81.18           65.68   
Johnson High School                    80.97           66.06   
Pena High School                       84.04           94.59   
Rodriguez High School                  80.74           66.37   
Shelton High School                    83.73           93.87   
Thomas High School                     83.85           93.27   
Wilson High School                     83.99           93.87   
Wright High School                     83.96           93.33   

                       % Passing Reading  % Overall Passing  
school_name                                                  
Bailey High School                 81.93              54.64  
Cabrera High School                97.04              91.33  
Figueroa High School               80.74              53.20  
Ford High School                   79.30              54.29  
Griffin High School                97.14              90.60  
Hernandez High School              80.86              53.53  
Holden High School                 96.25              89.23  
Huang High School                  81.32              53.51  
Johnson High School  

## Top Performing Schools (By % Overall Passing)

* Sort and display the top five performing schools by % overall passing.

In [10]:
#total passing top 5
top_by_passing = school_sum.sort_values("% Overall Passing", ascending=False)
top_by_passing.head(5)

Type  Total Students  Total School Budget  \
school_name                                                         
Cabrera High School  Charter            1858              1081356   
Thomas High School   Charter            1635              1043130   
Griffin High School  Charter            1468               917500   
Wilson High School   Charter            2283              1319574   
Pena High School     Charter             962               585858   

                     Per Student Budget  Average Math Score  \
school_name                                                   
Cabrera High School               582.0           83.061895   
Thomas High School                638.0           83.418349   
Griffin High School               625.0           83.351499   
Wilson High School                578.0           83.274201   
Pena High School                  609.0           83.839917   

                     Average Reading Score  % Passing Math  % Passing Reading  \
school_name                                                                     
Cabrera High School              83.975780           94.13              97.04   
Thomas High School               83.848930           93.27              97.31   
Griffin High School              83.816757           93.39              97.14   
Wilson High School               83.989488           93.87              96.54   
Pena High School                 84.044699           94.59              95.95   

                     % Overall Passing  
school_name                             
Cabrera High School              91.33  
Thomas High School               90.95  
Griffin High School              90.60  
Wilson High School               90.58  
Pena High School                 90.54

## Bottom Performing Schools (By % Overall Passing)

* Sort and display the five worst-performing schools by % overall passing.

In [11]:
#bottom passing top 5
bot_by_passing = school_sum.sort_values("% Overall Passing", ascending=True)
bot_by_passing.head(5)

Type  Total Students  Total School Budget  \
school_name                                                            
Rodriguez High School  District            3999              2547363   
Figueroa High School   District            2949              1884411   
Huang High School      District            2917              1910635   
Hernandez High School  District            4635              3022020   
Johnson High School    District            4761              3094650   

                       Per Student Budget  Average Math Score  \
school_name                                                     
Rodriguez High School               637.0           76.842711   
Figueroa High School                639.0           76.711767   
Huang High School                   655.0           76.629414   
Hernandez High School               652.0           77.289752   
Johnson High School                 650.0           77.072464   

                       Average Reading Score  % Passing Math  \
school_name                                                    
Rodriguez High School              80.744686           66.37   
Figueroa High School               81.158020           65.99   
Huang High School                  81.182722           65.68   
Hernandez High School              80.934412           66.75   
Johnson High School                80.966394           66.06   

                       % Passing Reading  % Overall Passing  
school_name                                                  
Rodriguez High School              80.22              52.99  
Figueroa High School               80.74              53.20  
Huang High School                  81.32              53.51  
Hernandez High School              80.86              53.53  
Johnson High School                81.22              53.54

## Math Scores by Grade

* Create a table that lists the average Reading Score for students of each grade level (9th, 10th, 11th, 12th) at each school.

  * Create a pandas series for each grade. Hint: use a conditional statement.
  
  * Group each series by school
  
  * Combine the series into a dataframe
  
  * Optional: give the displayed data cleaner formatting

In [12]:
maths_grades = school_data_complete.groupby(["grade", "school_name"]).mean()
maths_grades = maths_grades["math_score"]

maths_grades = pd.DataFrame({
    "9th": maths_grades["9th"],
    "10th": maths_grades["10th"],
    "11th": maths_grades["11th"],
    "12th": maths_grades["12th"],
})
maths_grades =round(maths_grades, 2)
maths_grades

9th   10th   11th   12th
school_name                                      
Bailey High School     77.08  77.00  77.52  76.49
Cabrera High School    83.09  83.15  82.77  83.28
Figueroa High School   76.40  76.54  76.88  77.15
Ford High School       77.36  77.67  76.92  76.18
Griffin High School    82.04  84.23  83.84  83.36
Hernandez High School  77.44  77.34  77.14  77.19
Holden High School     83.79  83.43  85.00  82.86
Huang High School      77.03  75.91  76.45  77.23
Johnson High School    77.19  76.69  77.49  76.86
Pena High School       83.63  83.37  84.33  84.12
Rodriguez High School  76.86  76.61  76.40  77.69
Shelton High School    83.42  82.92  83.38  83.78
Thomas High School     83.59  83.09  83.50  83.50
Wilson High School     83.09  83.72  83.20  83.04
Wright High School     83.26  84.01  83.84  83.64

## Reading Score by Grade 

* Perform the same operations as above for reading scores


In [13]:
reading_grades = school_data_complete.groupby(["grade", "school_name"]).mean()
reading_grades = reading_grades["reading_score"]

reading_grades = pd.DataFrame({
    "9th": reading_grades["9th"],
    "10th": reading_grades["10th"],
    "11th": reading_grades["11th"],
    "12th": reading_grades["12th"],
})
reading_grades =round(reading_grades, 2)
reading_grades

9th   10th   11th   12th
school_name                                      
Bailey High School     81.30  80.91  80.95  80.91
Cabrera High School    83.68  84.25  83.79  84.29
Figueroa High School   81.20  81.41  80.64  81.38
Ford High School       80.63  81.26  80.40  80.66
Griffin High School    83.37  83.71  84.29  84.01
Hernandez High School  80.87  80.66  81.40  80.86
Holden High School     83.68  83.32  83.82  84.70
Huang High School      81.29  81.51  81.42  80.31
Johnson High School    81.26  80.77  80.62  81.23
Pena High School       83.81  83.61  84.34  84.59
Rodriguez High School  80.99  80.63  80.86  80.38
Shelton High School    84.12  83.44  84.37  82.78
Thomas High School     83.73  84.25  83.59  83.83
Wilson High School     83.94  84.02  83.76  84.32
Wright High School     83.83  83.81  84.16  84.07

## Scores by School Spending

* Create a table that breaks down school performances based on average Spending Ranges (Per Student). Use 4 reasonable bins to group school spending. Include in the table each of the following:
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * Overall Passing Rate (Average of the above two)

In [14]:
# Create the bins in which Data will be held   
spend_bins = [0, 584.99, 629.99, 644.99, 680]

# Create the names for the five bins
spend_categs = ["<$585", "$585-630", "$630-645", "$645-680"]

#cut data into bins and summarise into new dataframe
school_sum["Spending Ranges (Per Student)"] = pd.cut(school_sum["Per Student Budget"], spend_bins, labels=spend_categs, include_lowest=True)
spend_summary = school_sum.groupby(["Spending Ranges (Per Student)"]).mean()

#trim back to relevant columns and display
spend_summary = spend_summary[[ 
      "Average Math Score", 
      "Average Reading Score",
      "% Passing Math",
      "% Passing Reading",
      "% Overall Passing"
     ]]
spend_summary = round(spend_summary, 2)
spend_summary 

Average Math Score  Average Reading Score  \
Spending Ranges (Per Student)                                              
<$585                                       83.46                  83.93   
$585-630                                    81.90                  83.16   
$630-645                                    78.52                  81.62   
$645-680                                    77.00                  81.03   

                               % Passing Math  % Passing Reading  \
Spending Ranges (Per Student)                                      
<$585                                   93.46              96.61   
$585-630                                87.13              92.72   
$630-645                                73.48              84.39   
$645-680                                66.16              81.13   

                               % Overall Passing  
Spending Ranges (Per Student)                     
<$585                                      90.37  
$585-630                                   81.42  
$630-645                                   62.86  
$645-680                                   53.53

## Scores by School Size

* Perform the same operations as above, based on school size.

In [15]:
# Create the bins in which Data will be held   
size_bins = [0, 999, 1999, 5000]

# Create the names for the five bins
size_categs = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000) "]

#cut data into bins and summarise into new dataframe
school_sum["School Size"] = pd.cut(school_sum["Total Students"], size_bins, labels=size_categs, include_lowest=True)
size_summary = school_sum.groupby(["School Size"]).mean()

#trim back to relevant columns and display
size_summary = size_summary[[ 
      "Average Math Score", 
      "Average Reading Score",
      "% Passing Math",
      "% Passing Reading",
      "% Overall Passing"
     ]]
size_summary = round(size_summary, 2)
size_summary


Average Math Score  Average Reading Score  % Passing Math  \
School Size                                                                     
Small (<1000)                    83.82                  83.93           93.55   
Medium (1000-2000)               83.37                  83.86           93.60   
Large (2000-5000)                77.75                  81.34           69.96   

                    % Passing Reading  % Overall Passing  
School Size                                               
Small (<1000)                   96.10              89.88  
Medium (1000-2000)              96.79              90.62  
Large (2000-5000)               82.77              58.28

## Scores by School Type

* Perform the same operations as above, based on school type

In [16]:
type_summary = school_sum.groupby(["Type"])

type_summary = type_summary[[
        "Total Students",
        "Per Student Budget",
        "Average Math Score", 
        "Average Reading Score",
        "% Passing Math",
        "% Passing Reading",
        "% Overall Passing"
     ]]

type_summary = round(type_summary.mean(), 2)

type_summary


#typemeans["% Passing Maths"] = 

#y = goodreaders.groupby(["type"])
#type_summary["% Passing Reading"] = round((y.count()["reading_score"]/school_sum["Total Students"] * 100), 2)

#z = goodboth.groupby(["type"])
#type_summary["% Overall Passing"] = round((z.count()["reading_score"]/school_sum["Total Students"] * 100), 2)


Total Students  Per Student Budget  Average Math Score  \
Type                                                               
Charter          1524.25              599.50               83.47   
District         3853.71              643.57               76.96   

          Average Reading Score  % Passing Math  % Passing Reading  \
Type                                                                 
Charter                   83.90           93.62              96.59   
District                  80.97           66.55              80.80   

          % Overall Passing  
Type                         
Charter               90.43  
District              53.67

## Conclusions


#1:
There is a huge difference in passing outcomes between Charter schools and District schools; according to these data, charter schools produce 90% of students passing reading and maths, compared with only 54% passing in District schools.

However this could be a correlation due to a third variable, such as the types of families who choose to send their children to these schools.

#2
Charter schools are on average much smaller than district schools. The data imply that students perform better in smaller schools.

#3: 
Interestingly, according to this data, charter schools have better educational outcomes than district schools despite a lower per-student budget. 

Perhaps having more flexibility in setting their own curriculum has enabled Charter schools to focus on only activities that contribute to on-paper success? only a broader measure of success than reading and maths will shed light on this.

(additional research drawn from https://www.usnews.com/education/k12/articles/understanding-charter-schools-vs-public-schools. Yes it's just the top result on google.)